cobbler/settings/migrations/V3_3_0.py

```diff
@@ -369,4 +382,87 @@ def migrate(settings: dict) -> dict:
     # delete removed keys
     helper.key_delete("cache_enabled", settings)
 
+    # migrate stored cobbler collections
+    migrate_cobbler_collections("/var/lib/cobbler/collections/")
+
     return normalize(settings)
+
+
+def backup_dir(dir_path: str):
+    """
+    Copies the directory tree and adds a suffix ".backup.XXXXXXXXX" to it.
+
+    :param dir_path: The full path to the directory which should be backed up.
+    :raises FileNotFoundError: In case the path specified was not existing.
+    """
+    src = os.path.normpath(dir_path)
+    now_iso = datetime.datetime.now().isoformat()
+    copytree(dir_path, f"{src}.backup.{now_iso}")
+
+
+def migrate_cobbler_collections(collections_dir: str):
+    """
+    Manipulate the main Cobbler stored collections and migrate deprecated settings
+    to work with newer Cobbler versions.
+
+    :param collections_dir: The directory of Cobbler where the collections files are.
+    """
+    backup_dir(collections_dir)
+    for collection_file in glob.glob(
+        os.path.join(collections_dir, "**/*.json"), recursive=True
+    ):
+        data = None
+        with open(collection_file, encoding="utf-8") as _f:
+            data = json.loads(_f.read())
+
+        # null values to empty strings
+        for key in data:
+            if data[key] is None:
+                data[key] = ""
+
+        # boot_loader -> boot_loaders
+        if "boot_loader" in data:
+            data["boot_loaders"] = data.pop("boot_loader")
+
+        # next_server -> next_server_v4, next_server_v6
+        if "next_server" in data:
+            addr = data["next_server"]
+            if addr == "<<inherit>>":
+                data["next_server_v4"] = addr
+                data["next_server_v6"] = addr
+                data.pop("next_server")
+            else:
+                _ip = ipaddress.ip_address(addr)
+                if isinstance(_ip, ipaddress.IPv4Address):
+                    data["next_server_v4"] = data.pop("next_server")
+                elif isinstance(_ip, ipaddress.IPv6Address):
+                    data["next_server_v6"] = data.pop("next_server")
```

Comment:
`next-server` can also be a hostname. For details see [here](https://serverfault.com/a/986804/408066) or in the isc DHCPd manpage.

Reply:
@SchoolGuy I've pushed an extra commit to take care of the case the `next_server` value of a collection is actually a hostname (expect an extra push to fix the spotted lint issues).

OTOH, looking at the cobbler settings, it seems `next_server` on 3.1.2 [1] was expecting to be IP and not hostname. Is this really correct?

If hostname is expected I think we would need to backport this to our Uyuni and SUMA package.

Reply:
After reverse-engineering the code and checking with 2.8.5 code it is visible that this was allowed to be a hostname in the past, however the stricter validation now only allows IP addresses. Thus this should be fine.